src/v/iceberg/tests/update_partition_spec_action_test.cc

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "iceberg/tests/test_schemas.h"
+#include "iceberg/transaction.h"
+
+#include <gtest/gtest.h>
+
+using namespace iceberg;
+
+class UpdatePartitionSpecActionTest : public ::testing::Test {
+public:
+    // Create a simple table with no snapshots.
+    table_metadata create_table() {
+        auto s = schema{
+          .schema_struct = std::get<struct_type>(test_nested_schema_type()),
+          .schema_id = schema::id_t{0},
+          .identifier_field_ids = {},
+        };
+        chunked_vector<schema> schemas;
+        schemas.emplace_back(s.copy());
+
+        chunked_vector<partition_spec> specs;
+        specs.push_back(partition_spec{.spec_id = partition_spec::id_t{0}});
+
+        return table_metadata{
+          .format_version = format_version::v2,
+          .table_uuid = uuid_t::create(),
+          .location = uri("s3://foo/bar"),
+          .last_sequence_number = sequence_number{0},
+          .last_updated_ms = model::timestamp::now(),
+          .last_column_id = s.highest_field_id().value(),
+          .schemas = std::move(schemas),
+          .current_schema_id = schema::id_t{0},
+          .partition_specs = std::move(specs),
+          .default_spec_id = partition_spec::id_t{0},
+          .last_partition_id = partition_field::id_t{999},
+        };
+    }
+};
+
+TEST_F(UpdatePartitionSpecActionTest, TestAddSpec) {
+    transaction tx(create_table());
+    ASSERT_EQ(tx.table().partition_specs.size(), 1);
+
+    using field = unresolved_partition_spec::field;
+    chunked_vector<field> fields{field{
+      .source_name = {"bar"},
+      .transform = bucket_transform{.n = 16},
+      .name = "field1",
+    }};
+
+    {
+        // add a simple spec
+        auto res = tx.set_partition_spec(unresolved_partition_spec{
+                                           .fields = fields.copy(),
+                                         })
+                     .get();
+        ASSERT_FALSE(res.has_error());
+        ASSERT_EQ(tx.table().partition_specs.size(), 2);
+        ASSERT_EQ(tx.table().default_spec_id, 1);
+        const auto& spec = tx.table().partition_specs.back();
```

Comment:
nitpick: the container of partition specs is an implementation detail, so it might be a bit more precise to lookup by spec ID